### PR TITLE
[GLUTEN-1372][clickhouse_backend] fix null partition read for ch

### PR DESCRIPTION
**Pull request by lhuang09287750**
## What changes were proposed in this pull request?

Spark+ch backend will read the null partition column as the default value of partition column type and
this behavior is differrent from spark which will read the null just as null.
So this pr add a fix to process null partition column read problem.

Fixes: \#1372

## How was this patch tested?

`  test("TPCDS Q99") {
    runTPCDSQuery("q99") { df => }
  }`
add this to GlutenClickHouseTPCDSParquetSuite to test

### CODE REVIEW

File: backends-clickhouse/src/test/scala/io/glutenproject/execution/GlutenClickHouseTPCDSParquetSuite.scala [hunk 1]
```diff
@@ -91,7 +91,7 @@ class GlutenClickHouseTPCDSParquetSuite extends GlutenClickHouseTPCDSAbstractSui
         |)
         |""".stripMargin
     val result = spark.sql(testSql).collect()
-    assert(result(0).getLong(0) == 791980)
+    assert(result(0).getLong(0) == 791809)
```
Comment (zzcclp) [inline backends-clickhouse/src/test/scala/io/glutenproject/execution/GlutenClickHouseTPCDSParquetSuite.scala:94]: the value 791809 is the same as the one of vanilla spark ?

File: cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.cpp [hunk 2]
```diff
@@ -240,6 +240,21 @@ DB::ColumnPtr FileReaderWrapper::createConstColumn(DB::DataTypePtr data_type, co
     return column;
 }
 
+DB::ColumnPtr FileReaderWrapper::createColumn(DB::DataTypePtr data_type, size_t rows, const String & value)
+{
+    if (StringUtils::isNullPartitionValue(value))
+    {
+        auto nested_type = DB::removeNullable(data_type);
+        auto column = nested_type->createColumnConstWithDefaultValue(rows);
+        return DB::ColumnNullable::create(column, DB::ColumnUInt8::create(rows, 1));
```
Comment (liuneng1994) [inline cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.cpp:249]: use makeNullable

File: backends-clickhouse/src/test/scala/io/glutenproject/execution/GlutenClickHouseTPCDSParquetSuite.scala (hunk 1 above)
Comment (lhuang09287750) [reply backends-clickhouse/src/test/scala/io/glutenproject/execution/GlutenClickHouseTPCDSParquetSuite.scala:94]: yes,791809 is the result of vanilla spark

File: cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.cpp [hunk 3]
```diff
@@ -240,6 +240,21 @@ DB::ColumnPtr FileReaderWrapper::createConstColumn(DB::DataTypePtr data_type, co
     return column;
 }
 
+DB::ColumnPtr FileReaderWrapper::createColumn(DB::DataTypePtr data_type, size_t rows, const String & value)
+{
+    if (StringUtils::isNullPartitionValue(value))
+    {
+        auto nested_type = DB::removeNullable(data_type);
```
Comment (liuneng1994) [inline cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.cpp:247]: 为什么不用getNestedType，这里是不是隐含一个假设,value是NullPartitionValue则datatype一定是Nullable。如果参数不满足假设会如何

File: cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.h [hunk 4]
```diff
@@ -29,6 +29,7 @@ class FileReaderWrapper
     FormatFilePtr file;
 
     static DB::ColumnPtr createConstColumn(DB::DataTypePtr type, const DB::Field & field, size_t rows);
+    static DB::ColumnPtr createColumn(DB::DataTypePtr data_type, size_t rows, const String & value);
```
Comment (liuneng1994) [inline cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.h:32]: 参数顺序建议与上面的createConstColumn保持一致，rows放到最后

File: cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.cpp (hunk 3 above)
Comment (lhuang09287750) [reply cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.cpp:247]: __HIVE_DEFAULT_PARTITION__是空列的默认值，也是在StringUtils::isNullPartitionValue里判断的，不过用户可以改这个默认值，的确会有风险，但没有好办法。为什么不用getNestedType--这个是什么

Comment (liuneng1994) [reply cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.cpp:247]: 需要加判断，如果不满足假设跑出logical error。
DataTypeNullable::getNestedType

File: cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.cpp (hunk 2 above)
Comment (lhuang09287750) [reply cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.cpp:249]: makeNullable seems not to be what i need.  It cannnot gen a totally null column which null map is all filled with 1

File: cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.cpp (hunk 3 above)
Comment (lhuang09287750) [reply cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.cpp:247]: 为什么不用getNestedType----在removeNullable里已经做了强制类型转换并调用了getNestedType

File: cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.h (hunk 4 above)
Comment (lhuang09287750) [reply cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.h:32]: done

File: cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.cpp (hunk 3 above)
Comment (liuneng1994) [reply cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.cpp:247]: 这里主要是因为你用了removeNullable这个函数可以接收任意类型，如果先判断datatype是否能转换为datatypeNullable，自然会出现一个else的情况需要处理

Comment (lhuang09287750) [reply cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.cpp:247]: DB::ColumnPtr FileReaderWrapper::createColumn(const String & value, DB::DataTypePtr type, size_t rows)
{
    if (StringUtils::isNullPartitionValue(value))
    {
        if (!type->isNullable())
        {
            throw DB::Exception(
                DB::ErrorCodes::LOGICAL_ERROR, "Partition column is null value,but column data type is not nullable.");
        }
        auto nested_type = static_cast<const DB::DataTypeNullable &>(*type).getNestedType();
        auto column = nested_type->createColumnConstWithDefaultValue(rows);
        return DB::ColumnNullable::create(column, DB::ColumnUInt8::create(rows, 1));
    }
    else
    {
        auto field = buildFieldFromString(value, type);
        return createConstColumn(type, field, rows);
    }
}
 what about this？

**Review by liuneng1994 — APPROVED** [review]
LGTM

**Pull request merged**